tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala

```diff
@@ -219,6 +219,221 @@ class TestHybrid extends BuilderTest {
+                builder.exists(
+                    builder.name(s"F$j", OperT1(paramTs_j, bodyT_j)),
+                    builder.bool(true),
+                ),
+                declParams.zipWithIndex.map { case ((bodyT, paramTs), k) =>
+                  builder.decl(
+                      s"F$k",
+                      builder.name(s"X$k", bodyT),
+                      paramTs.zipWithIndex.map { case (tt, i) => builder.param(s"p${i}_$k", tt) }: _*
+                  )
+                }: _*
+            ),
+            // LET  F1 == X1, ...,  Fj == \E Fj: TRUE, ..., FN = XN IN e, bound Fj has the right type
+            builder.letIn(
+                builder.name("e", letT),
+                declParams.zipWithIndex.map { case ((bodyT, paramTs), k) =>
+                  builder.decl(
+                      s"F$k",
+                      if (k == j) builder.exists(builder.name(s"F$j", OperT1(paramTs, bodyT)), builder.bool(true))
+                      else builder.name(s"X$k", bodyT),
+                      paramTs.zipWithIndex.map { case (tt, i) => builder.param(s"p${i}_$k", tt) }: _*
+                  )
+                }: _*
+            ),
+        )
+      }
+    }
+
+    def isExpected1(tparam: TParam1): TlaEx => Boolean = {
+      val (_, (bodyT, paramTs)) = tparam
+      val params = paramTs.zipWithIndex.map { case (tt, i) => builder.param(s"p$i", tt) }
+      val letT = OperT1(paramTs, bodyT)
+      val expectedBody =
+        LetInEx(
+            builder.name("F", letT),
+            builder.decl("F", builder.name("X", bodyT), params: _*),
+        )(Typed(letT))
+
+      expectedBody.eqTyped
+    }
+
+    def isExpectedN(tparam: TParamN): TlaEx => Boolean = {
+      val (letT, declParams) = tparam
+      val decls = declParams.zipWithIndex.map { case ((bodyT, paramTs), j) =>
+        val params = paramTs.zipWithIndex.map { case (tt, i) => builder.param(s"p${i}_$j", tt) }
+        builder.decl(s"F$j", builder.name(s"X$j", bodyT), params: _*)
+      }
+      val expectedBody =
+        decls.foldRight[TlaEx](builder.name("e", letT)) { case (decl, partial) =>
+          LetInEx(partial, decl)(Typed(letT))
+        }
+
+      expectedBody.eqTyped
+    }
+
+    def run[TParam](
+        mkWellTyped: TParam => T,
+        forceScopeException: TParam => Seq[T],
+        isExpected: TParam => TlaEx => Boolean,
+      )(tparam: TParam): Boolean = {
+
+      val goodDecl = mkWellTyped(tparam)
+
+      isExpected(tparam)(goodDecl) shouldBe true
+
+      val badDecls = forceScopeException(tparam)
+
+      badDecls.foreach { instruction =>
+        assertThrows[TBuilderScopeException] {
+          build(instruction)
+        }
+      }
+      true
+    }
+
+    checkRun[TParam1](run(mkWellTyped1, forceScopeException1, isExpected1))
+    checkRun[TParamN](run(mkWellTypedN, forceScopeExceptionN, isExpectedN))
+
+    // throws on n = 0
+    assertThrows[IllegalArgumentException] {
+      builder.letIn(builder.int(0))
+    }
+
+  }
```

Comment:
This is a very big test, imo, and i cannot  tell by reading this part what properties or behaviors are actually being tested. I know that the test is testing `let-in`, but other than than I can't tell what kind of things are mean to hold here. 

I suggest either refactoring the test into a few distinct property tests that clearly communicate and exemplify what is being tested or adding comments here that explain what is being tested in each chunk. The former is much more preferable, IMO, but I'd be fine with just the latter.

Reply:
I've split the tests and added comments, most notably to the method `runLetTests`.